### ISSUE TITLE
Retransmit data on folder rename

### ISSUE THREAD

**Issue by jabozksl** [syncthing/syncthing#9230]
Renaming folders makes syncthing to retransmit all of its content to other machines.
I thought it could just to rename a folder on clients instead of that.
Version used: v1.26.0, (windows and linux)

**Comment by hireworksltd** [comment]
This has been mentioned on the forum recently... great post below:

https://forum.syncthing.net/t/file-renames-moves/20463/11

Would the first option in the above post be a possible folder config addition?
We would absolutely love this feature and would be happy to fund any work towards it.

Merry Christmas/Seasons Greeting to you all and congrats on 10 years of Syncthing! It's a truly awesome piece of work.

Best,
JL

**Comment by calmh** [comment]
Yeah, I think it's less about creating an option for it and more about doing it to begin with. :) As it is mostly about timing, this is something that could be handled as mentioned (keep all the changes private until renames etc have been "mapped", and then send in the proper order). That's not trivial, though.

**Comment by calmh** [comment]
Currently, a scan works in two phases:

1. look for new/changed content and hash it
2. look for deleted files

While scanning, changes are grouped together in batches and then committed to the database, which in turns triggers the change to be sent to connected peers.

The receiving side will react to incoming changes by starting a "pull", which looks at the set of changes known at that point and tries to match up adds and deletes into renames.

The end result of this is that a single renamed file, for example, is pretty much guaranteed to have the add+delete end up in the same batch and properly recognised as a rename on the receiving side. However a large rename (a directory with many files in it) will likely generate at least one full batch of just additions and only later a batch of deletes, and these may well end up being handled separately on the receiving side: no optimisation.

What we could possibly do is be smarter in the scanning stage: when we index a new file, see if we already had an existing file with the same block hashes in the database. If we do, check if that file still exists. If it doesn't, record the delete immediately. We should then end up with adds+deletes in close proximity and have a very large probability of applying them as a rename.

**Comment by AudriusButkevicius** [comment]
Isn't this implemented already?

https://github.com/syncthing/syncthing/blob/acd767b30b818c77b79bc439a377213e91e55c48/lib/model/folder.go#L858

As it processes items coming out the scan, it finds items with the same BlocksHash, and checks if it's still there. If not, puts the add and delete one after another, and forces them to be in the same batch.

**Comment by calmh** [comment]
🤦 So it is, but I didn't find it despite looking right at it. So then I don't know what the problem is, it should just work.

**Comment by hireworksltd** [comment]
Hi all,

Did a bit of testing over the holidays on this. Running latest version on all devices it used the rename shortcut every time when renaming a large folder (100+ items). We've a bit of a strange use case that I believe might be causing this issue on our setups in particular. I'm not sure how this would be fixable without a way to defer sending any changes until the folder is idle (and without errors) after a rescan.

The issue in our case seems to be when renaming a directory when a scan is already in progress (ie. a directory is renamed, a scan starts, and then the directory is renamed again). The syncthing folder obviously errors out as expected with multiple not found errors and schedules another scan, but the deletion events cause the files to be deleted remotely (presumably because there is no corresponding file created event) a the files re-transferred. 

I may be reading that wrong from my cursory understanding of how these renames work, apologies if so!

We use Syncthing to keep Avid Media Composer media in sync, which requires that media is indexed, which (annoyingly) involves multiple folder renames to get media online.

Best,
JL

**Comment by taokann** [comment]
Hi,
Just experienced this problem today.
I renamed a 1GB folder and now 1GB is being sent over the network.
Before that the folder was idle.

Syncthing is v1.27.2-ds4 on Debian 13.
